Bug in C++ codegen for set stringify when setting size

### Description

Increasing the size of a set using the size pragma causes invalid C++ code output.
```nim
type
    Flags* = enum
        A
        B
    Mask* {.size: sizeof(uint16).} = set[Flags]

proc check(flags: Mask) =
    echo flags

check {A, B}
```
Using a size of 1 byte works, but anything larger does not. C output works regardless of size.

### Nim Version

Nim Compiler Version 2.1.1 [Linux: amd64]
Compiled at 2024-06-02
Copyright (c) 2006-2024 by Andreas Rumpf

git hash: 4bd1cf2376a25b14a28103db639707bbd99a44d0
active boot switches: -d:release

### Current Output

```text
/usr/bin/ld: ~/.cache/nim/b_d/@mb.nim.cpp.o: in function `check__b_u8(unsigned long)':
@mb.nim.cpp:(.text+0x1d8): undefined reference to `dollar___b_u95(unsigned long)'
/usr/bin/ld: .../b: hidden symbol `_Z14dollar___b_u95m' isn't defined
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
```


### Expected Output

```text
{A, B}
```


### Possible Solution

_No response_

### Additional Information

Tested 1.6.20, 2.0.0 and 2.0.4 using the Nim playground and none of them work.

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -1793,7 +1793,7 @@ proc typeSectionFinalPass(c: PContext, n: PNode) =
         if pragmas[i].kind == nkExprColonExpr and
             pragmas[i][0].kind == nkIdent and
             whichKeyword(pragmas[i][0].ident) == wSize:
-          if s.typ.kind != tyEnum and sfImportC notin s.flags:
+          if s.typ.kind != tyEnum and sfImportc notin s.flags:
             # EventType* {.size: sizeof(uint32).} = enum
             # AtomicFlag* {.importc: "atomic_flag", header: "<stdatomic.h>", size: 1.} = object
             localError(c.config, pragmas[i].info, "size pragma only allowed for enum types and imported types")

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -1796,7 +1796,7 @@ proc typeSectionFinalPass(c: PContext, n: PNode) =
           if s.typ.kind != tyEnum and sfImportC notin s.flags:
             # EventType* {.size: sizeof(uint32).} = enum
             # AtomicFlag* {.importc: "atomic_flag", header: "<stdatomic.h>", size: 1.} = object
-            localError(c.config, pragmas[i].info, "size pragma only allowed for imported objects and enums")
+            localError(c.config, pragmas[i].info, "size pragma only allowed for enum types and imported types")
 
     if a[1].kind == nkEmpty:
       var x = a[2]

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -1787,6 +1787,17 @@ proc typeSectionFinalPass(c: PContext, n: PNode) =
     # check the style here after the pragmas have been processed:
     styleCheckDef(c, s)
     # compute the type's size and check for illegal recursions:
+    if a[0].kind == nkPragmaExpr:
+      let pragmas = a[0][1]
+      for i in 0 ..< pragmas.len:
+        if pragmas[i].kind == nkExprColonExpr and
+            pragmas[i][0].kind == nkIdent and
+            whichKeyword(pragmas[i][0].ident) == wSize:
+          if s.typ.kind notin {tyObject, tyEnum}:
+            # EventType* {.size: sizeof(uint32).} = enum
+            # AtomicFlag* {.importc: "atomic_flag", header: "<stdatomic.h>", size: 1.} = object
+            localError(c.config, pragmas[i].info, "size pragma only allowed for objects and enums")
+
     if a[1].kind == nkEmpty:
       var x = a[2]
       if x.kind in nkCallKinds and nfSem in x.flags:

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -1793,10 +1793,10 @@ proc typeSectionFinalPass(c: PContext, n: PNode) =
         if pragmas[i].kind == nkExprColonExpr and
             pragmas[i][0].kind == nkIdent and
             whichKeyword(pragmas[i][0].ident) == wSize:
-          if s.typ.kind notin {tyObject, tyEnum}:
+          if s.typ.kind != tyEnum and sfImportC notin s.flags:
             # EventType* {.size: sizeof(uint32).} = enum
             # AtomicFlag* {.importc: "atomic_flag", header: "<stdatomic.h>", size: 1.} = object
-            localError(c.config, pragmas[i].info, "size pragma only allowed for objects and enums")
+            localError(c.config, pragmas[i].info, "size pragma only allowed for imported objects and enums")
 
     if a[1].kind == nkEmpty:
       var x = a[2]

